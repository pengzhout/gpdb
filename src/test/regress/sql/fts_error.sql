create extension if not exists gp_inject_fault;
select count(*) = 2 as in_sync from gp_segment_configuration
where content = 0 and mode = 's';
-- Once this fault is hit, FTS process should abort current
-- transaction and exit.
select gp_inject_fault('fts_update_config', 'error', '', '', '', -1, 0, 1);
-- FTS probe connection should encounter an error due to this fault,
-- injected on content 0 primary.
select gp_inject_fault('fts_handle_message', 'error', '', '', '', -1, 0, dbid)
from gp_segment_configuration where content = 0 and role = 'p';
-- Upon failure to probe content 0 primary, FTS will try to update the
-- configuration.  The update to configuration will hit error due to
-- the "fts_update_config" fault.
select gp_wait_until_triggered_fault('fts_update_config', 1, 1);

select gp_inject_fault('fts_handle_message', 'reset', dbid)
from gp_segment_configuration where content = 0 and role = 'p';

select gp_inject_fault('fts_update_config', 'reset', 1);

-- Verify that FTS didn't leak any locks due to the error during
-- config update.
select locktype, mode, relation, pid, granted from pg_locks where
relation = 'gp_segment_configuration'::regclass or
relation = 'gp_configuration_history'::regclass;
select gp_inject_fault('fts_update_config', 'reset', 1);
-- Postmaster should have restarted FTS by now.  Trigger a scan and
-- validate that configuration is sane.
select gp_request_fts_probe_scan();
select count(*) = 2 as in_sync from gp_segment_configuration
where content = 0 and mode = 's';
