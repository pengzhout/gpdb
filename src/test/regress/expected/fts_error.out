create extension if not exists gp_inject_fault;
select count(*) = 2 as in_sync from gp_segment_configuration
where content = 0 and mode = 's';
 in_sync 
---------
 t
(1 row)

-- Once this fault is hit, FTS process should abort current
-- transaction and exit.
select gp_inject_fault('fts_update_config', 'error', '', '', '', -1, 0, 1);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

-- FTS probe connection should encounter an error due to this fault,
-- injected on content 0 primary.
select gp_inject_fault('fts_handle_message', 'error', '', '', '', -1, 0, dbid)
from gp_segment_configuration where content = 0 and role = 'p';
NOTICE:  Success:  (seg0 127.0.1.1:25432 pid=15439)
 gp_inject_fault 
-----------------
 t
(1 row)

-- Upon failure to probe content 0 primary, FTS will try to update the
-- configuration.  The update to configuration will hit error due to
-- the "fts_update_config" fault.
select gp_wait_until_triggered_fault('fts_update_config', 1, 1);
NOTICE:  Success:
 gp_wait_until_triggered_fault 
-------------------------------
 t
(1 row)

select gp_inject_fault('fts_handle_message', 'reset', dbid)
from gp_segment_configuration where content = 0 and role = 'p';
NOTICE:  Success:  (seg0 127.0.1.1:25432 pid=15439)
 gp_inject_fault 
-----------------
 t
(1 row)

select gp_inject_fault('fts_update_config', 'reset', 1);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

-- Verify that FTS didn't leak any locks due to the error during
-- config update.
select locktype, mode, relation, pid, granted from pg_locks where
relation = 'gp_segment_configuration'::regclass or
relation = 'gp_configuration_history'::regclass;
 locktype | mode | relation | pid | granted 
----------+------+----------+-----+---------
(0 rows)

select gp_inject_fault('fts_update_config', 'reset', 1);
NOTICE:  Success:
 gp_inject_fault 
-----------------
 t
(1 row)

-- Postmaster should have restarted FTS by now.  Trigger a scan and
-- validate that configuration is sane.
select gp_request_fts_probe_scan();
 gp_request_fts_probe_scan 
---------------------------
 t
(1 row)

select count(*) = 2 as in_sync from gp_segment_configuration
where content = 0 and mode = 's';
 in_sync 
---------
 t
(1 row)

